### DISCUSSION TITLE
Exporting to .CSV from PowerBI date and time format

### DISCUSSION THREAD

**Discussion by Bayesd12** [DaxStudio/DaxStudio#1037] [General]
I am having a issue with the way the Date and Time is formatting when I export to .CSV. 
This is how my data looks in PowerBI
![Dax studio fix 2](https://user-images.githubusercontent.com/123781209/215223169-f8d43004-adaa-4917-8f26-8eb2412ed329.png)
This is how my data Exports to .CSV
![Dax studio fix](https://user-images.githubusercontent.com/123781209/215278692-4fa4dd54-be13-43c1-99d6-7f41dfc89bfd.png)
The date column has to be manually formatted to get it back to it original state, This is a time consuming process due to the number of sheets that will be exported at a time. 
![Dax studio fix 3](https://user-images.githubusercontent.com/123781209/215226621-1feb9f8e-81d4-4333-81f4-3e8a11cc816e.png)
The time column requires a lot more editing to make it usable as a .CSV file
![Dax studio fix](https://user-images.githubusercontent.com/123781209/215223256-eb78f7b3-cf80-4948-8180-c433413ff97b.png)

What can I do to resolve this issue.
Dax Studio is a great tool, any help to work around the issue is appreciated.

**Comment by Bayesd12** [comment]
I have looked at the associated discussions where this problem has already been raised and I have looked at all the relevant posts on a google search, what is apparent is there hasn't been a published solution to the formatting issue. 
 
Can  a DAX Query be built to fix the format upon exporting?? Any idea appreciated

**Comment by dgosbell** [comment]
What you are looking at here is not the csv output - it is Excel's interpretation of the csv content. To see the actual csv output you would need to use a text editor like notepad or VS Code.

However we actually write the query results to the file output without formatting by design. Numbers are written out with their full precision and datetimes columns are written out using an ISO 8601 format. This is because this option writes text based files, so there is no facility to include metadata about the column's data type or formatting.

Datetimes can be particularly problematic as the different regional formats mean that it can be possible to write a file on one computer that either cannot be parsed on a computer with different regional settings or in a worst case scenario you can get "corruption" where some records may get the day and month swapped without error leading to incorrect results.

The file output is designed to work as a way of getting accurate data out of the Power BI data model. It is not designed as an alternative to building a report. If you have specific formatting requirements one option would be to copy your query to a paginated report and apply your required formatting there. Using the Excel output is another option as that file format has richer capabilities in terms of data types and formatting on top of the raw data. Or you could potentially create measures using the FORMAT function to force a specific format to be used.

**Reply by Bayesd12** [reply]
Thank you for your detailed reply, I have got to work on the format function  and have generated a query to achieve my objective as far as "Date" & "Time" columns are concerned.
I am having a issue with the "Order By" part of the query. 
What I am finding is that the Date is not been chronologically ordered. 
The output calculates on the 1st of the month from years 1991 -2023 and then it calculates all the 2nd of the month from 1991-2023 and so on. How can I get the query to recognise the chronological order of the date column
**Please see by below Dax Query**
EVALUATE
SELECTCOLUMNS('_01st_7th', "DATE",FORMAT([DATE],"dd/mm/YYYY"),
"time",FORMAT([time],"hh:mm"),
"OPEN",(_01st_7th[OPEN]),
"HIGH",(_01st_7th[HIGH]),
"LOW",(_01st_7th[LOW]),
"CLOSE",(_01st_7th[CLOSE]))

ORDER BY
[DATE] ASC,[TIME] 

**Please see below a image of how the data is looking**   
![dax studio Order by 1](https://user-images.githubusercontent.com/123781209/217789102-28b4a58e-4389-42c7-aaf7-5434b28613b9.png)

**Reply by dgosbell** [reply]
This is happening because FORMAT returns a string value so the sorting is being done alphabetically. And you can only sort by columns that are available in the output so one option would be to add a second date column that is not formatted then order by that, but that does muck up your output.

Why is sorting and formatting important for you in a csv file? 

The csv format has been designed for data transfer which is why it uses an unambiguous format for datetime columns. 

What sort of system are you loading this data into? 

Or if it is being read by a person then have to tried using the xlsx format instead?

**Comment by Bayesd12** [comment]
dgosbell, Thanks for reviewing my issue.
The software I'm using is reading .CSV files into as Date/time/open/high/low/close data and the format is important to it being able to read it. What other approach do you think may be suitable given my requirement to sort the output into chronological order? 
I have seen references to using measures to format the date & time. I haven't able to get the query to work though. In your experience would it be a work around and something I should pursue? thanks

**Reply by Bayesd12** [reply]
Anyone able to help me progress beyond this point. Is there a better way to achieve the format and be able to sort the Date Column correctly

**Reply by dgosbell** [reply]
So, if you can't build a paginated report for this you could try this preview build of the next release https://ci.appveyor.com/project/DarrenGosbell/daxstudio/builds/46260316/artifacts - I've added a new output option called "Formatted File" which will output to a text-based format but will use the same formatting logic as the data grid. I've only just started testing this, but it appears to be working OK.

**Reply by Bayesd12** [reply]
dgosbell, Thank you for your help with this, appreciate it. Where do I add this log to Dax Studio to allow me to try this preview build? Thanks

**Reply by dgosbell** [reply]
There is a setup.exe at the link I posted above.

**Reply by Bayesd12** [reply]
thank you... it was right under my nose ;-)

**Reply by dgosbell** [reply]
Actually the .zip portable version is there too if you want to test this without overwriting the current release... :)

**Comment by Bayesd12** [comment]
I'm still having difficulty getting the date column to format, I have copied in my query. What am i missing here??

EVALUATE
SELECTCOLUMNS('_2100_0800_01st_07th', "DATE",FORMAT([DATE],"DD/MM/YYYY"),
"time",FORMAT([time],"hh:mm"),
"OPEN",(_2100_0800_01st_07th[OPEN]),
"HIGH",(_2100_0800_01st_07th[HIGH]),
"LOW",(_2100_0800_01st_07th[LOW]),
"CLOSE",(_2100_0800_01st_07th[CLOSE]))
ORDER BY
[DATE] ASC, [TIME] ASC 

![Dax studio 22nd feb](https://user-images.githubusercontent.com/123781209/220765181-d1a5b6db-9fb6-4378-bd70-264055d12f8b.png)

**Reply by dgosbell** [reply]
What difficulty are you having?? Is it the ordering? If so, that is because you are still manually formatting the data into strings still.

If you just output the columns (making sure the output names have the same names as the columns in your data model the the new formatted file output should apply the formatstrings as it writes the csv file.
![image](https://user-images.githubusercontent.com/8241830/221073421-9f1077d3-7b9c-4f69-a9d3-118f01de9deb.png)

**Comment by Bayesd12** [comment]
You are a Legend, thank you for assisting me with this. Top man

**Comment by Bayesd12** [comment]
Ok, Dax studio is working a treat at exporting the correct format to .CSV... what a fantastic tool. 

Here's a image of the Query I am running for multiple .CSV files.
![Dax studio how to aaply name](https://user-images.githubusercontent.com/123781209/221554347-fb26b4e9-800a-4fa3-ae58-4426f415b592.png)

Here how Dax shows it in the log/results/history 
![Dax studio how to aaply name2](https://user-images.githubusercontent.com/123781209/221554528-3439e9b1-47e3-4ac8-9408-9fa841b4c862.png)

And here is how it saves it to a document...using numbers to identify each log
![Dax studio how to aaply name3](https://user-images.githubusercontent.com/123781209/221554612-1ba67546-155d-4c00-9e40-5e4b87aa85e0.png)

Its working beautifully as far as the data goes, I would also like it to use the name which I have circled in blue in this final image when it saves the CSV. 
![Dax studio how to aaply name2b](https://user-images.githubusercontent.com/123781209/221555823-2ad632d1-7c89-4427-91f6-6e77acdb50e8.png)
What can i add to my query to include the name when it saves the new .CSV to document.

**Reply by Bayesd12** [reply]
Any chance these is a solution to saving the files using their files names when it outputs to Excel?

**Reply by dgosbell** [reply]
If you use the data export in the Advanced tab to export an entire table I output to tabs / files with the names of the tables in the data model. 

But if you have a set of arbitrary queries there is no easy way to figure out what the files should be called. I have a design I've been working on for using specially formatted comments to allow for setting the file names, but I don't even have a prototype yet, I've just been working on the specifications.
